test/check-machines-networks

```diff
@@ -795,6 +795,72 @@ class TestMachinesNetworks(VirtualMachinesCase):
         self.waitNetworkRow("test_network2", connectionName, False)
 
     def testNetworkAddStaticDCHPHosts(self):
+        class NetworkAddStaticDCHPHostsDialog(object):
+            def __init__(
+                self, test_obj=None, network_name="default", connection_name="system", mac=None, ip=None, entry_id=0,
+                xfail=None, remove=True
+            ):
+                self.test_obj = test_obj
+                self.network_name = network_name
+                self.connection_name = connection_name
+                self.mac = mac
+                self.ip = ip
+                self.entry_id = entry_id
+                self.xfail = xfail
+                self.remove = remove
+
+            def execute(self):
+                self.open()
+                self.fill()
+                self.add()
+                if not self.xfail:
+                    self.verify_dialog()
+                    self.verify_overview()
+                    if self.remove:
+                        self.cleanup()
+
+            def open(self):
+                b.click(f"#network-{self.network_name}-{self.connection_name}-static-host-entries-add")
+                b.wait_visible("#add-new-static-entry-mac-address")
+                b.wait_in_text(".pf-c-modal-box .pf-c-modal-box__header .pf-c-modal-box__title", "Add a DHCP static host entry")
+
+            def fill(self):
+                b.set_input_text("#add-new-static-entry-mac-address", self.mac)
+                b.set_input_text("#add-new-static-entry-ip-address", self.ip)
+
+            def cancel(self):
+                b.click(".pf-c-modal-box__footer button:contains(Cancel)")
+                b.wait_not_present("#add-new-static-entry")
+
+            def add(self):
+                b.click(".pf-c-modal-box__footer button:contains(Add)")
+
+                if (self.xfail):
+                    # Check incomplete dialog
+                    b.wait_in_text(".pf-c-modal-box .pf-c-alert.pf-m-danger", self.xfail)
+                    self.cancel()
+                else:
+                    b.wait_not_present("#add-new-static-entry")
+
+            def verify_dialog(self):
+                # Check that the defined network is now visible
+                b.wait_in_text("body", "Networks")
+
+                # Verify libvirt XML
+                net_xml = f"virsh -c qemu:///system net-dumpxml {self.network_name}"
+                xmllint_element = f"{net_xml} | xmllint --xpath 'string(//network/ip/dhcp/host/{{prop}})' - 2>&1 || true"
```

Comment:
Why would this fail legitimately? I really don't like alternative truths in tests, they should be predictable. Also, it's usually better to keep stderr as-is, so that errors appear in the log instead of being swallowed into the string comparison. Can you please drop this? If there is a case where the XML does not have a dhcp/host/prop path (the plausible way to fail this), it's fine if the test fails here already?

Reply:
Yes, it's totally normal that virtual network doesn't have whole <dhcp> elemenent subtree, and even if it has, it could also not have <dhcp> subtree. 
If in the future, ours (or libvirt's) code for adding DHCP hosts regresses in a way that dhcp hosts it would not be present in the XML, then with ` - 2>&1 || true` the error message would look like this:
```
+ virsh -c qemu:///system net-dumpxml default | xmllint --xpath 'string(//network/ip/dhcp/host/@mac)' - 2>&1 || true

Traceback (most recent call last):
  File "/usr/lib64/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 887, in testNetworkAddStaticDCHPHosts
    ).execute()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 817, in execute
    self.verify_dialog()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 853, in verify_dialog
    self.test_obj.assertEqual(self.mac, m.execute(xmllint_element.format(prop='@mac')).strip())
  File "/usr/lib64/python3.10/unittest/case.py", line 845, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python3.10/unittest/case.py", line 1226, in assertMultiLineEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/usr/lib64/python3.10/unittest/case.py", line 675, in fail
    raise self.failureException(msg)
AssertionError: '52:54:00:6b:70:af' != ''
- 52:54:00:6b:70:af
+ 

```

Without `- 2>&1 || true` the error message would look like this:

```
+ virsh -c qemu:///system net-dumpxml default | xmllint --xpath 'string(//network/ip/dhcp/host/@mac)'
Usage : xmllint [options] XMLfiles ...
	Parse the XML files and output the result of the parsing
	--version : display the version of the XML library used
	--debug : dump a debug tree of the in-memory document
	--shell : run a navigating shell
	--debugent : debug the entities defined in the document
	--copy : used to test the internal copy implementation
	--recover : output what was parsable on broken XML documents
	--huge : remove any internal arbitrary parser limits
	--noent : substitute entity references by their value
	--noenc : ignore any encoding specified inside the document
	--noout : don't output the result tree
	--path 'paths': provide a set of paths for resources
	--load-trace : print trace of all external entities loaded
	--nonet : refuse to fetch DTDs or entities over network
	--nocompact : do not generate compact text nodes
	--htmlout : output results as HTML
	--nowrap : do not put HTML doc wrapper
	--valid : validate the document in addition to std well-formed check
	--postvalid : do a posteriori validation, i.e after parsing
	--dtdvalid URL : do a posteriori validation against a given DTD
	--dtdvalidfpi FPI : same but name the DTD with a Public Identifier
	--quiet : be quiet when succeeded
	--timing : print some timings
	--output file or -o file: save to a given file
	--repeat : repeat 100 times, for timing or profiling
	--insert : ad-hoc test for valid insertions
	--compress : turn on gzip compression of output
	--html : use the HTML parser
	--xmlout : force to use the XML serializer when using --html
	--nodefdtd : do not default HTML doctype
	--push : use the push mode of the parser
	--pushsmall : use the push mode of the parser using tiny increments
	--memory : parse from memory
	--maxmem nbbytes : limits memory allocation to nbbytes bytes
	--nowarning : do not emit warnings from parser/validator
	--noblanks : drop (ignorable?) blanks spaces
	--nocdata : replace cdata section with text nodes
	--format : reformat/reindent the output
	--encode encoding : output in the given encoding
	--dropdtd : remove the DOCTYPE of the input docs
	--pretty STYLE : pretty-print in a particular style
	                 0 Do not pretty print
	                 1 Format the XML content, as --format
	                 2 Add whitespace inside tags, preserving content
	--c14n : save in W3C canonical format v1.0 (with comments)
	--c14n11 : save in W3C canonical format v1.1 (with comments)
	--exc-c14n : save in W3C exclusive canonical format (with comments)
	--nsclean : remove redundant namespace declarations
	--testIO : test user I/O support
	--catalogs : use SGML catalogs from $SGML_CATALOG_FILES
	             otherwise XML Catalogs starting from 
	         file:///etc/xml/catalog are activated by default
	--nocatalogs: deactivate all catalogs
	--auto : generate a small doc on the fly
	--xinclude : do XInclude processing
	--noxincludenode : same but do not generate XInclude nodes
	--nofixup-base-uris : do not fixup xml:base uris
	--loaddtd : fetch external DTD
	--dtdattr : loaddtd + populate the tree with inherited attributes 
	--stream : use the streaming interface to process very large files
	--walker : create a reader and walk though the resulting doc
	--pattern pattern_value : test the pattern support
	--chkregister : verify the node registration code
	--relaxng schema : do RelaxNG validation against the schema
	--schema schema : do validation against the WXS schema
	--schematron schema : do validation against a schematron
	--sax1: use the old SAX1 interfaces for processing
	--sax: do not build a tree but work just at the SAX level
	--oldxml10: use XML-1.0 parsing rules before the 5th edition
	--xpath expr: evaluate the XPath expression, imply --noout

Libxml project home page: https://gitlab.gnome.org/GNOME/libxml2
Traceback (most recent call last):
  File "/usr/lib64/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 887, in testNetworkAddStaticDCHPHosts
    ).execute()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 817, in execute
    self.verify_dialog()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 853, in verify_dialog
    self.test_obj.assertEqual(self.mac, m.execute(xmllint_element.format(prop='@mac')).strip())
  File "/home/skobyda/Work/machines/bots/machine/machine_core/ssh_connection.py", line 368, in execute
    raise subprocess.CalledProcessError(proc.returncode, command, output=output)
subprocess.CalledProcessError: Command 'virsh -c qemu:///system net-dumpxml default | xmllint --xpath 'string(//network/ip/dhcp/host/@mac)'' returned non-zero exit status 1.
```

Reply:
The `- 2>&1 || true` is used all over our place, and I think it was probably used to get clearer error message from assertions instead of just Command returned non-zero exit. Both error message are don't tell you where is the problem though, so I personally don't mind removing it

Reply:
Urgh, that's the worst conceivable output from xmllint for a "pattern not found" situation.. Thanks, nevermind then.

Reply:
In the end, I didn't really like error messages the both approaches were generating, so I adjusted the code a bit to create this output:
```
Traceback (most recent call last):
  File "/usr/lib64/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 892, in testNetworkAddStaticDCHPHosts
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 818, in execute
    self.verify_dialog()
  File "/home/skobyda/Work/machines/./test/check-machines-networks", line 859, in verify_dialog
Exception: No DHCP host with IP address 128.132.123.1 found in network XML:
<network>
  <name>default</name>
  <uuid>4c3eb707-770b-4d8e-b000-fabec9cff59e</uuid>
  <forward mode='nat'>
    <nat>
      <port start='1024' end='65535'/>
    </nat>
  </forward>
  <bridge name='virbr0' stp='on' delay='0'/>
  <mac address='52:54:00:0e:e7:7b'/>
  <ip address='192.168.122.1' netmask='255.255.255.0'>
    <dhcp>
      <range start='192.168.122.2' end='192.168.122.254'/>
      <host mac='52:54:00:e6:6b:9b' ip='192.168.122.222'/>
    </dhcp>
  </ip>
</network>
```

Reply:
Thanks! Indeed that's more explicit.